lib/Core/PTreeWriter.cpp

```diff
@@ -0,0 +1,169 @@
+//===-- PTreeWriter.cpp ---------------------------------------------------===//
+//
+//                     The KLEE Symbolic Virtual Machine
+//
+// This file is distributed under the University of Illinois Open Source
+// License. See LICENSE.TXT for details.
+//
+//===----------------------------------------------------------------------===//
+
+#include "PTreeWriter.h"
+
+#include "PTree.h"
+#include "klee/Support/ErrorHandling.h"
+
+using namespace klee;
+
+
+void prepare_statement(sqlite3 * db, std::string &query, sqlite3_stmt ** stmt) {
+#ifdef SQLITE_PREPARE_PERSISTENT
+  if (sqlite3_prepare_v3(db, query.c_str(), -1, SQLITE_PREPARE_PERSISTENT, stmt, nullptr) != SQLITE_OK) {
+#else
+  if (sqlite3_prepare_v3(db, query.c_str(), -1, 0, stmt, nullptr) != SQLITE_OK) {
+#endif
+    klee_warning("Process tree database: can't prepare query: %s [%s]", sqlite3_errmsg(db), query.c_str());
+    sqlite3_close(db);
+    klee_error("Process tree database: can't prepare query: %s", query.c_str());
```

Comment:
Why are you so careful about closing the database here? If you can't prepare statements you haven't written anything useful in the DB yet. Also having both a warning and an error with no controlflow inbetween, you should manaully format the message.

Reply:
I like to shutdown it gracefully (in case it already has content) and hence in `klee_error` I can't use `sqlite_errmsg`.

Reply:
Sure, but you can build the string before closing the db